docs/IR.md

```diff
@@ -1,78 +1,210 @@
 
 ### Graphs
 
-Each computation dataflow graph is structured as a list of nodes that form a graph. Each node is a call to an operator. Nodes have zero or more inputs, one or more outputs, and zero or more attribute-value pairs. Additionally, a valid ONNX graph must fulfill the following requirements:
+A serialized graph is comprised of a set of metadata fields, a list of model parameters, and a list of computation nodes.
+
+Each computation dataflow graph is structured as a topologically sorted list of nodes that form a graph, which MUST be free of cycles. Each node represents a call to an operator. Each node has zero or more inputs and one or more outputs.
+
+Graphs have the following properties:
+
+|Name|Type|Description|
+|---|---|---|
+name|string|The name of the model graph.
+node|Node[]|A list of nodes, forming a partially ordered computation graph based on input/output data dependencies.
+initializer|Tensor[]|A list of named tensor values, used to specify default values for some of the inputs of the graph. Each initializer value is associated with an input by name matching.
+doc_string|string|A human-readable documentation for this model. Markdown is allowed.
+input|ValueInfo[]|The input “parameters” of the graph, possibly initialized by a default value found in ‘initializer.’
+output|ValueInfo[]|The output parameters of the graph. Once all output parameters have been written to by a graph execution, the execution is complete.
+value_info|ValueInfo|__TODO: Write this description__
 
-- The graph MUST be free of cycles.
-- The graph MUST be in SSA (static single assignment) form, meaning outputs of all nodes are unique.
-- The nodes list MUST be in topologically sorted order, meaning if input of node `N2` is the output of node `N1`, `N2` must appear after `N1` in the nodes list.
 
-#### Model Graph Metadata
+Each graph MUST define the names and types of its inputs and outputs, which are specified as ‘value info’ structures, having the following properties:
 
-The following describes the metadata properties of a model graph:
+Name|Type|Description
+|---|---|---|
+name|string|The name of the value/parameter.
+type|Type|The type of the value.
+doc_string|string|A human-readable documentation for this value. Markdown is allowed.
 
-|Name|Type|Format|Description|
-|----|----|------|-----------|
-|name|string|Valid C identifier|A name for the model.|
-|doc_string|string|Free form|A human-readable documentation string intended to summarize the purpose of the model. Markdown is allowed.|
+Each graph MUST specify a name and a domain. Domains MUST be specified using reverse domain names as organization identifiers, the same convention that is used for naming Java packages.
```

Comment:
domain is used in opset and model too, but this naming requirement is only expressed here. Does this imply the other places do not have this requirement? Or should this requirement be moved up to apply broadly to any domain field?

Reply:
I don't think operator sets require the domain. Specifically, I believe that the default operator set leaves out the domain name.

That aside, my text is actually completely wrong -- graphs don't have a domain property, at all. The domain is on the model, and required. :-) Thanks for catching it!

Reply:
though it may be optional on opset, when it is specified it is expected to follow the format you describe.

Reply:
Got it. I thought your point was about the first sentence there. Agreed, all domains should follow the same reverse-DNS scheme.